src/leiningen/ring/war.clj

```diff
@@ -132,13 +132,27 @@
     (generate-resolve handler-sym)))
 
 (defn compile-servlet [project]
-  (let [servlet-ns  (symbol (servlet-ns project))]
+  (let [servlet-ns  (symbol (servlet-ns project))
+        init-sym    (get-in project [:ring :servlet-init])
+        destroy-sym (get-in project [:ring :servlet-destroy])]
     (compile-form project servlet-ns
       `(do (ns ~servlet-ns
-             (:gen-class :extends javax.servlet.http.HttpServlet))
+             (:gen-class
+               :extends javax.servlet.http.HttpServlet
+               :exposes-methods {~'init    ~'superInit
+                                 ~'destroy ~'superDestroy})
+             (:import javax.servlet.http.HttpServlet))
            (def ~'service-method)
            (defn ~'-service [servlet# request# response#]
-             (~'service-method servlet# request# response#)))
+             (~'service-method servlet# request# response#))
+           (defn ~'-init [this# cfg#]
+             ~(if init-sym
```

Comment:
Should this `if` be around the `-init` and `-destroy` functions? We don't need the functions at all if the symbol is `nil`.

Reply:
I had the exact same thought, but it's not 100% clear to me whether it will work, since the `init` and `destroy` methods are now _exposed_ by the generated class. If you're confident that `gen-class` doesn't care about not-exposing the listed as exposed methods, then I can have a go at your suggestion, otherwise I will have to test it against a real project.

Reply:
Probably should be:

```clojure
~@(if init-sym
    `((defn ~'-init [this# cfg#]
        (~(generate-resolve init-sym))
        (. this# (~'superInit cfg#)))))
```

Reply:
Hi again, and apologies for the delay - last week was kind of hectic for me... 

Ok, so feel free to check the new implementation. it does  what you asked, and it fixes a little bug with the previous version (i.e. calling `superInit` after self). What follows is a little test straight on the REPL:

```clj
;; dummies
(def my-init #(println "INIT"))
(def my-destroy #(println "DESTROY"))

(let [servlet-ns  'foo.bar
      init-sym    nil
      destroy-sym nil]
;; SNIP - the syntax-quoted expression from lines 139 - 157 
 )

```
=>
```clj 
(do
 (clojure.core/ns
  foo.bar
  (:gen-class :extends javax.servlet.http.HttpServlet :exposes-methods {init superInit, destroy superDestroy})
  (:import javax.servlet.http.HttpServlet))
 (def service-method)
 (clojure.core/defn -service
  [servlet__2073__auto__ request__2074__auto__ response__2075__auto__]
  (service-method servlet__2073__auto__ request__2074__auto__ response__2075__auto__))
 nil  ;; no init emitted
 nil) ;; no destroy emitted
```
Compare the above with the below (when `init-sym` and `destroy-sym` are non-nil)

```clj
(do
 (clojure.core/ns
  foo.bar
  (:gen-class :extends javax.servlet.http.HttpServlet :exposes-methods {init superInit, destroy superDestroy})
  (:import javax.servlet.http.HttpServlet))
 (def service-method)
 (clojure.core/defn -service
  [servlet__2299__auto__ request__2300__auto__ response__2301__auto__]
  (service-method servlet__2299__auto__ request__2300__auto__ response__2301__auto__))
 (clojure.core/defn -init
  [this__2296__auto__ cfg__2297__auto__]
  (. this__2296__auto__ (superInit cfg__2297__auto__))
  ((do (clojure.core/require (quote leiningen.ring.war)) (clojure.core/resolve (quote leiningen.ring.war/my-init)))))
 (clojure.core/defn -destroy
  [this__2298__auto__]
  ((do (clojure.core/require (quote leiningen.ring.war)) (clojure.core/resolve (quote leiningen.ring.war/my-destroy))))
  (. this__2298__auto__ superDestroy)))
```
So from the looks of it, we should be fine...I will follow this up with an confirmation  after an actual run.

Reply:
BTW, if I use plain `resolve` instead of `generate-resolve`, i get something much nicer:

```clj
(do
 (clojure.core/ns
  foo.bar
  (:gen-class :extends javax.servlet.http.HttpServlet :exposes-methods {init superInit, destroy superDestroy})
  (:import javax.servlet.http.HttpServlet))
 (def service-method)
 (clojure.core/defn -service
  [servlet__2387__auto__ request__2388__auto__ response__2389__auto__]
  (service-method servlet__2387__auto__ request__2388__auto__ response__2389__auto__))
 (clojure.core/defn -init
  [this__2384__auto__ cfg__2385__auto__]
  (. this__2384__auto__ (superInit cfg__2385__auto__))
  (#'leiningen.ring.war/my-init)) ;; <=========
 (clojure.core/defn -destroy 
  [this__2386__auto__] 
   (#'leiningen.ring.war/my-destroy) ;; <=========
   (. this__2386__auto__ superDestroy)))
```

Reply:
I can confirm that the code works. On the topic of `resolve` VS `generate-resolve`, I think what we want is `clojure.core/requiring-resolve`, but that is kind of new (added on 1.10) so I'm not sure how you feel about it.